pkg/proxy/conntrack/cleanup.go

```diff
@@ -27,85 +27,96 @@ import (
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/klog/v2"
 	"k8s.io/kubernetes/pkg/proxy"
-	proxyutil "k8s.io/kubernetes/pkg/proxy/util"
 	netutils "k8s.io/utils/net"
 )
 
-// CleanStaleEntries takes care of flushing stale conntrack entries for services and endpoints.
-func CleanStaleEntries(ct Interface, ipFamily v1.IPFamily, svcPortMap proxy.ServicePortMap,
-	serviceUpdateResult proxy.UpdateServiceMapResult, endpointsUpdateResult proxy.UpdateEndpointsMapResult) {
-	deleteStaleServiceConntrackEntries(ct, ipFamily, svcPortMap, serviceUpdateResult, endpointsUpdateResult)
-	deleteStaleEndpointConntrackEntries(ct, ipFamily, svcPortMap, endpointsUpdateResult)
-}
+// CleanStaleEntries evaluates all UDP flow entries and cleans up the ones which will
+// blackhole the traffic.
+func CleanStaleEntries(ct Interface, ipFamily v1.IPFamily,
+	svcPortMap proxy.ServicePortMap, endpointsMap proxy.EndpointsMap) {
 
-// deleteStaleServiceConntrackEntries takes care of flushing stale conntrack entries related
-// to UDP Service IPs. When a service has no endpoints and we drop traffic to it, conntrack
-// may create "black hole" entries for that IP+port. When the service gets endpoints we
-// need to delete those entries so further traffic doesn't get dropped.
-func deleteStaleServiceConntrackEntries(ct Interface, ipFamily v1.IPFamily, svcPortMap proxy.ServicePortMap, serviceUpdateResult proxy.UpdateServiceMapResult, endpointsUpdateResult proxy.UpdateEndpointsMapResult) {
-	var filters []netlink.CustomConntrackFilter
-	conntrackCleanupServiceIPs := serviceUpdateResult.DeletedUDPClusterIPs
-	conntrackCleanupServiceNodePorts := sets.New[int]()
-
-	// merge newly active services gathered from endpointsUpdateResult
-	// a UDP service that changes from 0 to non-0 endpoints is newly active.
-	for _, svcPortName := range endpointsUpdateResult.NewlyActiveUDPServices {
-		if svcInfo, ok := svcPortMap[svcPortName]; ok {
-			klog.V(4).InfoS("Newly-active UDP service may have stale conntrack entries", "servicePortName", svcPortName)
-			conntrackCleanupServiceIPs.Insert(svcInfo.ClusterIP().String())
-			for _, extIP := range svcInfo.ExternalIPs() {
-				conntrackCleanupServiceIPs.Insert(extIP.String())
-			}
-			for _, lbIP := range svcInfo.LoadBalancerVIPs() {
-				conntrackCleanupServiceIPs.Insert(lbIP.String())
-			}
-			nodePort := svcInfo.NodePort()
-			if svcInfo.Protocol() == v1.ProtocolUDP && nodePort != 0 {
-				conntrackCleanupServiceNodePorts.Insert(nodePort)
-			}
-		}
+	entries, err := ct.ListEntries(ipFamilyMap[ipFamily])
```

Comment:
I'm interested in logging the performance of this loop

```
start := time.Now()
klog.V(4).Infof("started to reconcile conntrack entries for IP family %s", ipFamily)
defer func() {
    klog.V(4).Infof("finished reconciling %d conntrack entries for IP Family %s took %v", len(entries), ipFamily, time.Since(start))
}()
```

Reply:
I'll try to compare the time taken here reconciler+netlink with old approach which involved os/exec of binary.

Reply:
I tried to compare cleanup time of reconciler/netlink with the previous os/exec based cleanup by running multiple e2e tests focusing only on conntrack tests.

Cleanup using reconciler + netlink:
```
finished reconciling conntrack entries for IP Family IPv4 took 9.569447ms
finished reconciling conntrack entries for IP Family IPv4 took 5.020591ms
finished reconciling conntrack entries for IP Family IPv4 took 12.588683ms
finished reconciling conntrack entries for IP Family IPv4 took 16.194495ms
finished reconciling conntrack entries for IP Family IPv4 took 24.310555ms
finished reconciling conntrack entries for IP Family IPv4 took 32.359003ms
finished reconciling conntrack entries for IP Family IPv4 took 10.588969ms
finished reconciling conntrack entries for IP Family IPv4 took 7.481758ms
finished reconciling conntrack entries for IP Family IPv4 took 29.750414ms
finished reconciling conntrack entries for IP Family IPv4 took 6.959896ms
finished reconciling conntrack entries for IP Family IPv4 took 22.832781ms
finished reconciling conntrack entries for IP Family IPv4 took 14.622177ms
finished reconciling conntrack entries for IP Family IPv4 took 4.890278ms
finished reconciling conntrack entries for IP Family IPv4 took 10.28007ms
finished reconciling conntrack entries for IP Family IPv4 took 6.691969ms
finished reconciling conntrack entries for IP Family IPv4 took 78.0888ms
finished reconciling conntrack entries for IP Family IPv4 took 12.340849ms
finished reconciling conntrack entries for IP Family IPv4 took 35.156297ms
finished reconciling conntrack entries for IP Family IPv4 took 10.072606ms
```

Cleanup using  os/exec
```
finished reconciling conntrack entries for IP Family IPv4 took 3.471µs
finished reconciling conntrack entries for IP Family IPv4 took 23.704773ms
finished reconciling conntrack entries for IP Family IPv4 took 129.793245ms
finished reconciling conntrack entries for IP Family IPv4 took 104.067371ms
finished reconciling conntrack entries for IP Family IPv4 took 112.93273ms
finished reconciling conntrack entries for IP Family IPv4 took 104.291622ms
finished reconciling conntrack entries for IP Family IPv4 took 46.103779ms
finished reconciling conntrack entries for IP Family IPv4 took 66.254913ms
finished reconciling conntrack entries for IP Family IPv4 took 147.694227ms
finished reconciling conntrack entries for IP Family IPv4 took 170.545757ms
finished reconciling conntrack entries for IP Family IPv4 took 39.512888ms
finished reconciling conntrack entries for IP Family IPv4 took 255.934412ms
finished reconciling conntrack entries for IP Family IPv4 took 54.708817ms
finished reconciling conntrack entries for IP Family IPv4 took 25.562314ms
finished reconciling conntrack entries for IP Family IPv4 took 160.29402ms
finished reconciling conntrack entries for IP Family IPv4 took 149.999816ms
finished reconciling conntrack entries for IP Family IPv4 took 88.143605ms
finished reconciling conntrack entries for IP Family IPv4 took 32.694975ms
finished reconciling conntrack entries for IP Family IPv4 took 3.714µs
```

Note: With the old approach when there are no `NewlyActiveUDPServices` and no `DeletedUDPEndpoints` cleanup is no-op (happens in <5µs).  With reconciler, there is always a netlink call to dump the table.
Also with the previous approach, the cleanup time will scale linearly with the number of flows to clean as a separate process is forked, with reconciler we only dump the table once and do the cleanup.

Reply:
well, if we don't have any udp service or all the udp services don't have any endpoint we can return faster, no?

`if len(ipIndex) == 0 || len `

Reply:
kube-dns will always exist in every cluster, right?
If 5ms latency really matters then we need to maintain states and conditionally do the cleanup.

Reply:
> kube-dns will always exist in every cluster, right?

no, it is an implementation of each cluster, you don't need to implement dns as a service ... but it is a good point, it is the most commonly deployed way

Reply:
I can use `endpointsUpdateResult` to figure out if any UDP service endpoints changed and reconcile acc.

Reply:
> well, if we don't have any udp service or all the udp services don't have any endpoint we can return faster, no?

No, because if a UDP endpoint was just deleted we want to delete the NAT flows for that endpoint.

> If 5ms latency really matters then we need to maintain states and conditionally do the cleanup.

5ms doesn't sound bad, but presumably the amount of time actually scales with the size of the conntrack table; and the frequency with which we need to do a sync scales with the number of service updates, and both of those scale with the size of the cluster. So there's the possibility that the amount of time taken could be overwhelming in large clusters.

Netlink lets you keep a connection open and receive notifications from the kernel when things change, like with a kube apiserver watch, so in theory we shouldn't need to download the state from scratch every time. Although, if there are _lots_ of conntrack changes, then we might spend more time keeping up with the diffs than we would spend just rereading the whole current state every time...

Anyway, yes, it would probably be good to make use of the information we have about whether _relevant_ service changes occurred. (Though also we want to be able to handle to startup case where we just reconcile without knowing what the previous state was.)

Reply:
> Anyway, yes, it would probably be good to make use of the information we have about whether _relevant_ service changes occurred.

(Though that doesn't mean we need to keep _all_ of the information that's currently in the map update results; we might be able to have just a single boolean?)

Reply:
> 5ms doesn't sound bad, but presumably the amount of time actually scales with the size of the conntrack table; and the frequency with which we need to do a sync scales with the number of service updates, and both of those scale with the size of the cluster. So there's the possibility that the amount of time taken could be overwhelming in large clusters.

we may implement something as you did with the rules, only do full reconciliation every X or when  we detect some change that requires it

> Netlink lets you keep a connection open and receive notifications from the kernel when things change, like with a kube apiserver watch, so in theory we shouldn't need to download the state from scratch every time. Although, if there are lots of conntrack changes, then we might spend more time keeping up with the diffs than we would spend just rereading the whole current state every time..

that implies we need to  build the same mechanism in the client watching the netlink changes that we have on the informers, and AFAIK netlink does not send version state and these capabilities th

Reply:
It will be super useful to know the implication of dumping the conntrack table periodically , if is negligible is ok, otherwise I wonder if we should feature gate it, to not roll this out and figure out several releases later that is a problem on very large clusters or with a lot of conntrack entries

Reply:
I have refactored the `EndpointsUpdateResult` to track if any UDP service changed endpoints, and am calling cleanup only when required.

Reply:
what is the difference then with what we have today?
Will this solve the existing gaps:
- cleanup during startup (can not find the link to this one)
- consider endpoint topology https://github.com/kubernetes/kubernetes/issues/126934
- deal with reoncilation failures https://github.com/kubernetes/kubernetes/issues/112604

Reply:
Ah, we actually can't rely on `EndpointsUpdateResult`. Have reverted the change. 

The current reconciler logic is to clear all the conntrack entries which have the Service as original destination and the reply source doesn't  match any of the serving endpoint IP.

```
 ( --orig-dst == ServiceIP | --orig-port-dst == ServiceNodePort ) && --reply-src != ServingEndpointIP
```
> cleanup during startup (can not find the link to this one)

UDP flows which are DNATed to a serving endpoint will be preserved during startup.


 > consider endpoint topology
 
I can't reproduce the issue locally (tested on kindest/node:v1.31.0).  Even if there are blackhole or stale entries in conntrack table with the new logic they should be cleared up immediately after the sync.

> deal with reoncilation failures

Since we don't rely on `EndpointsUpdateResult` for the required conntrack cleanup this issue will be solved with the new logic.

Reply:
> I'm interested in logging the performance of this loop

FWIW you can use `"k8s.io/utils/trace".LogIfLong()` to log only if something takes too long

Reply:
> Ah, we actually can't rely on `EndpointsUpdateResult`. Have reverted the change.

why?

Reply:
I added a field `ConntrackCleanupRequired` in `EndpointsUpdateResult` which is set to true if any UDP service has endpoints change. If in a particular sync `ConntrackCleanupRequired` is true and network-programming fails the conntrack cleanup won't be triggered, and if the next sync has no change in UDP services then the conntrack entries which we wanted to remove will never be removed. 


I feel like relying on `EndpointsUpdateResult` goes against the general reconciliation logic making it edge-triggered.

Reply:
asked in the netfilter mailing list for advice https://lore.kernel.org/netfilter/CABhP=tZe2MDYJWqmyhb=we2g_v11wuDb6yBp878vt7qQYrvUiA@mail.gmail.com/T/#u , let's wait for the experts answer